pkg/volume/metadata/metadata.go

```diff
@@ -0,0 +1,246 @@
+	return plugin.newCleanerInternal(volName, podUID, mounter)
+}
+
+func (plugin *metadataPlugin) newCleanerInternal(volName string, podUID types.UID, mounter mount.Interface) (volume.Cleaner, error) {
+	return &metadataVolume{volName: volName,
+		podRef:  &api.ObjectReference{UID: podUID},
+		plugin:  plugin,
+		mounter: mounter}, nil
+}
+
+// metadataVolume handles retrieving metadata from the API server
+// and placing them into the volume on the host.
+type metadataVolume struct {
+	volName                string
+	fieldReferenceFileName map[string]string
+	podRef                 *api.ObjectReference
+	plugin                 *metadataPlugin
+	opts                   *volume.VolumeOptions
+	mounter                mount.Interface
+}
+
+func (m *metadataVolume) SetUp() error {
+	return m.SetUpAt(m.GetPath())
+}
+
+// This is the spec for the volume that this plugin wraps.
+var wrappedVolumeSpec = &volume.Spec{
+	Name:         "not-used",
+	VolumeSource: api.VolumeSource{EmptyDir: &api.EmptyDirVolumeSource{Medium: api.StorageTypeMemory}},
+}
+
+// readFile reads full path file name and return it as a string
+func readFile(fileName string) string {
+	data, err := ioutil.ReadFile(fileName)
+	if err != nil {
+		fmt.Printf("%s\n", err)
+		return ""
+	}
+	return string(data)
+}
+
+// generateTmpFile writes input parameter 'values' in a temporary file. File descriptor
+// is returned. In case error is returned file descriptor is not meaningful
+func generateTmpFile(outputDir, values string) (*os.File, error) {
+	prefix := "tmpfile"
+	tmpFile, err := ioutil.TempFile(outputDir, prefix)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create tmp file in dir '%s', with prefix '%s'", outputDir, prefix)
+	}
+	err = ioutil.WriteFile(tmpFile.Name(), []byte(values), 0644)
+	if err != nil {
+		return nil, err
+	}
+	return tmpFile, nil
+}
+
+// isFileToGenerate compares actual file with the new values. If
+// different (or the file does not exist) return true,
+func isFileToGenerate(outputDir, fileName, values string) bool {
+	return readFile(path.Join(outputDir, fileName)) != values
+}
+
+// writeInFile writes string in fileName file if string must be updated.
+// It writes 'atomically' writing in a tmp file and re-placing the previous copy
+// via symbol linking recreation. It ends up removing the previous file
+func (m *metadataVolume) writeInFile(fileName, values string) error {
+	if values == "" || fileName == "" {
+		return nil
+	}
+	if !isFileToGenerate(m.GetPath(), fileName, values) {
+		return nil
+	}
+	fd, err := generateTmpFile(m.GetPath(), values)
+	if err != nil {
+		fmt.Printf("%s\n", err)
+		return err
+	}
+
+	// unlink the file
+	finalFilePath := path.Join(m.GetPath(), fileName)
+	actualTarget, e := os.Readlink(finalFilePath)
+	if e != nil {
+		// no link so creates one
+		os.Symlink(fd.Name(), finalFilePath)
+		return nil
+	}
+	e = os.Remove(finalFilePath) // remove the link
```

Comment:
I don't think "atomic" means the same thing to you and I :)

Remove and link is *not* atomic.  The only atomic filesyste op is rename.  Write the new tmp file, os.Rename() it to the name you want.